### PR TITLE
Reset commands

### PR DESCRIPTION
**Pull request by oanesbogdan**
- bbapp:update 
     - adjusted to accept host/port/user/password as options
     - drop all tables
- bbapp:create_site
    - create a new website entry in site table
    - creates the layouts and the root page
- bbapp:create_sudoer
    - creates a new sudoer
- users:update_rights
    - updates the user group rights
    - posibility to create a new user and assign a usergroup

### CODE REVIEW

File: Console/Command/ApplicationUpdateCommand.php [hunk 1]
```diff
@@ -70,14 +76,97 @@ protected function configure()
...
+            $connection['user'] = $input->getOption('user');
+        }
+
+        if (null !== $input->getOption('password')) {
+            $connection['password'] = $input->getOption('password');
+        }
+
+        $environment = $input->getOption('env');
+
+        $bbapp = new \BackBee\Standard\Application(null, $environment);
```
Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: A bbapp is aleready created with environment and context options:
$bbapp = $this->getContainer()->get('bbapp');

File: Console/Command/ApplicationUpdateCommand.php [hunk 2]
```diff
@@ -70,14 +76,97 @@ protected function configure()
...
+            $doctrine_config['dbal']['proxy_ns'] = 'Proxies';
+        }
+
+        if (!array_key_exists('proxy_dir', $doctrine_config['dbal'])) {
+            $doctrine_config['dbal']['proxy_dir'] = $bbapp->getCacheDir() . '/' . 'Proxies';
+        }
+
+        try {
+            $em = \BackBee\Util\Doctrine\EntityManagerCreator::create($doctrine_config['dbal']);
+
```
Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: Could you please add a Use instruction

File: Console/Command/ApplicationUpdateCommand.php [hunk 3]
```diff
@@ -70,14 +76,97 @@ protected function configure()
...
+                    ->connect('master');
+            } else {
+                $em->getConnection()
+                    ->connect();
+            }
+
+        } catch (\Exception $e) {
+            throw new \BackBee\Exception\DatabaseConnectionException(
+                'Unable to connect to the database.', 0, $e);
+        }
```
Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: Could you please add a Use instruction

File: Console/Command/ApplicationUpdateCommand.php [hunk 4]
```diff
@@ -70,14 +76,97 @@ protected function configure()
...
+        // Overwrite doctrine.yml file with new configuration
+        if(isset($connection['user']) && isset($connection['password'])) {
+            unset($doctrine_config['dbal']['proxy_ns']);
+            unset($doctrine_config['dbal']['proxy_dir']);
+            file_put_contents(
+                $bbapp->getBaseDir().DIRECTORY_SEPARATOR.'repository/Config/doctrine.yml',
+                Yaml::dump($doctrine_config)
+            );
+            $output->writeln('<info>Doctrine.yml updated with new configuration.</info>');
+        }
```
Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: I think there is no need to update doctrine.yml file here

File: Console/Command/ApplicationUpdateCommand.php [hunk 5]
```diff
@@ -70,14 +76,97 @@ protected function configure()
...
+
+    /**
      * {@inheritdoc}
      */
     protected function execute(InputInterface $input, OutputInterface $output)
     {
         $force = $input->getOption('force');
+        $drop = $input->getOption('drop');
 
         $bbapp = $this->getContainer()->get('bbapp');
```
Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: $bbapp become unused

File: Console/Command/ApplicationUpdateCommand.php [hunk 6]
```diff
@@ -107,20 +196,30 @@ protected function execute(InputInterface $input, OutputInterface $output)
...
+                $sqls = $this->getUpdateQueries();
+                $schema->updateSchema($metadata, true);
+
+                $output->writeln('<info>SQL executed: </info>'.PHP_EOL.implode(";".PHP_EOL, $sqls).'');
+            }
             $this->em->getConnection()->executeUpdate('SET FOREIGN_KEY_CHECKS=1');
         }
-
-        $output->writeln('<info>SQL executed: </info>'.PHP_EOL.implode(";".PHP_EOL, $sqls).'');
     }
```
Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: without force option, the command should output queries to be run.
So I would keep:
```
$sqls = $this->getUpdateQueries();
if ($drop) {
    $sqls = $schema->getDropDatabaseSQL() . PHP_EOL . $sqls;
}

if ($force) {
    //execute dropDatabase(); and updateSchema($metadata, true);
}

$output->writeln('<info>SQL executed: </info>'.PHP_EOL.implode(";".PHP_EOL, $sqls).'');
```

File: Console/Command/CreateSudoerCommand.php [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    /** 
+     * Init main tools
+     */
+    protected function init(InputInterface $input, OutputInterface $output)
+    {
+    	$this->input = $input;
+    	$this->output = $output;
+
+    	$this->app = $this->getApplication()->getApplication();
+    	$this->entyMgr = $this->app->getEntityManager();
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: $input, $output, $app and $entyMgr must be declared as private

File: Console/Command/CreateSudoerCommand.php [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+        # Check if user already exists in db
+        if ($this->entyMgr->getRepository('BackBee\Security\User')->findOneBy(array('_login' => $this->input->getOption('user_name')))) {
+            $this->output->writeln('<info>User is already present in DB.</info>');
+            return $this;
+        }
+
+        $environment = $this->input->getOption('env');
+
+        $bbapp = new \BackBee\Standard\Application(null, $environment);
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: application is already available in $this->app

File: Console/Command/CreateSudoerCommand.php [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        if (!$this->input->getOption('user_name') || !$this->input->getOption('user_password') || !$this->input->getOption('user_email')) {
+            $this->output->writeln('<info>You have to specify all option in order to insert a new user (--user_name, --user_password, --user_email)</info>');
+            return $this;
+        }
+
+        # Check if user already exists in db
+        if ($this->entyMgr->getRepository('BackBee\Security\User')->findOneBy(array('_login' => $this->input->getOption('user_name')))) {
+            $this->output->writeln('<info>User is already present in DB.</info>');
+            return $this;
+        }
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: We can want to add an existing user as sudoer, so if $user exists, add it in security.yml elsewhere create a new user

File: Console/Command/CreateSudoerCommand.php [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+
+        $encoder = $encoderFactory->getEncoder($adminUser);
+        $adminUser
+            ->setPassword($encoder->encodePassword($this->input->getOption('user_password'), ''))
+            ->setEmail($this->input->getOption('user_email'))
+            ->generateRandomApiKey()
+        ;
+
+        $this->entyMgr->persist($adminUser);
+        $this->entyMgr->flush($adminUser);
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: From lines 109 to 130 should be a private method createUser($login, $password, $email) returning a Backbee\Security\User

File: Console/Command/CreateSudoerCommand.php [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+        # Recreate security.yml
+		$securityConf = $this->app->getConfig()->getSecurityConfig();
+        if (!array_key_exists($adminUser->getLogin(), $securityConf['sudoers'])) {
+            $securityConf['sudoers'][$adminUser->getLogin()] = $adminUser->getId();
+        }
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: if user is already a sudoer simply return, don't update security.yml file

File: Console/Command/CreateSudoerCommand.php [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+		
+		file_put_contents(
+			$bbapp->getBaseDir().DIRECTORY_SEPARATOR.'repository/Config/security.yml', 
+			Yaml::dump($securityConf)
+		);
+
+        $this->output->writeln('<info>New user created.</info>');
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: From lines 132 to 141 should be a private methode addSudoer(User $user) returning TRUE or FALSE

File: Console/Command/CreateSudoerCommand.php [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+    /** 
+     * Execute command
+     */
+    protected function execute(InputInterface $input, OutputInterface $output)
+    {
+    	$this
+    		->init($input, $output)
+    		->runInsertSudoerProcess()
+		;
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: There are too tabs

File: Console/Command/CreateSudoerCommand.php [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+            ->setPassword($encoder->encodePassword($this->input->getOption('user_password'), ''))
+            ->setEmail($this->input->getOption('user_email'))
+            ->generateRandomApiKey()
+        ;
+
+        $this->entyMgr->persist($adminUser);
+        $this->entyMgr->flush($adminUser);
+
+        # Recreate security.yml
+		$securityConf = $this->app->getConfig()->getSecurityConfig();
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: One too tab

File: Console/Command/CreateSudoerCommand.php [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+        if (!array_key_exists($adminUser->getLogin(), $securityConf['sudoers'])) {
+            $securityConf['sudoers'][$adminUser->getLogin()] = $adminUser->getId();
+        }
+		
+		file_put_contents(
+			$bbapp->getBaseDir().DIRECTORY_SEPARATOR.'repository/Config/security.yml', 
+			Yaml::dump($securityConf)
+		);
```
Comment (crouillon) [inline Console/Command/CreateSudoerCommand.php]: Could you format thes lines

File: Console/Command/CreateWebsiteCommand.php [hunk 16]
```diff
@@ -0,0 +1,227 @@
...
+    	$this->output = $output;
+
+    	$this->app = $this->getApplication()->getApplication();
+    	$this->entyMgr = $this->app->getEntityManager();
+
+    	$this->siteLabel = $this->input->getOption('site_label');
+
+		if (!filter_var($this->input->getOption('site_domain'), FILTER_VALIDATE_URL)){
+			$this->output->writeln('<error>Invalid site domain format. Example of valid domain: http://backbee.com</error>');
+		}
```
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: Could you format these lines

File: Console/Command/CreateWebsiteCommand.php [hunk 17]
```diff
@@ -0,0 +1,227 @@
...
+
+		if (is_array($sitesConf) && array_key_exists($this->siteLabel, $sitesConf)) {
+			$this->output->writeln('<info>This label already present in sites.yml</info>'); 
+			return $this;
+		}
+
+		if (is_array($sitesConf) && in_array($this->siteDomain, array_column($sitesConf, 'domain'))) {
+			$this->output->writeln('<info>This domain already present in sites.yml</info>'); 
+			return $this;
+		}
```
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: Format please

File: Console/Command/CreateWebsiteCommand.php [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        ];		
+
+        $newSitesConf = (is_array($sitesConf)) ? array_merge($sitesConf, $newSite) : $newSite;
+
+		file_put_contents(
+			$this->app->getBaseDir().DIRECTORY_SEPARATOR.'repository/Config/sites.yml', 
+			Yaml::dump($newSitesConf)
+		);
+
+		return $this;
```
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: Format please

File: Console/Command/CreateWebsiteCommand.php [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+            $blockDemo = new \BackBee\ClassContent\BlockDemo();
+            $blockDemo->setState(\BackBee\ClassContent\AbstractClassContent::STATE_NORMAL);
+            $blockDemo->setRevision(1);
+
+            $homeContainer = new \BackBee\ClassContent\Home\HomeContainer();
+            $homeContainer->setState(\BackBee\ClassContent\AbstractClassContent::STATE_NORMAL);
+            $homeContainer->setRevision(1);
+            $homeContainer->container->setState(\BackBee\ClassContent\AbstractClassContent::STATE_NORMAL);
+            $homeContainer->container->setRevision(1);
+            $homeContainer->container->push($blockDemo);
```
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: Being in Core PHP, we can invoque BackBee Standard class contents, the simpliest way to fix it is to remove lines 211 to 220

File: Console/Command/CreateWebsiteCommand.php [hunk 20]
```diff
@@ -0,0 +1,227 @@
...
+                ->setLabel('Category')
+                ->setPicPath($categoryLayout->getUid() . '.png')
+                ->setSite($site)
+            ;
+            $this->entyMgr->persist($categoryLayout);
+        }
```
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: Keep only the Home layout creation please

File: Console/Command/CreateWebsiteCommand.php [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+            $homeContainer = new \BackBee\ClassContent\Home\HomeContainer();
+            $homeContainer->setState(\BackBee\ClassContent\AbstractClassContent::STATE_NORMAL);
+            $homeContainer->setRevision(1);
+            $homeContainer->container->setState(\BackBee\ClassContent\AbstractClassContent::STATE_NORMAL);
+            $homeContainer->container->setRevision(1);
+            $homeContainer->container->push($blockDemo);
+
+            $page->getContentSet()->first()->push($homeContainer);
+        }        
+
```
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: I think that the mediacenter root creation is missing:
```
                // Creating mediacenter root
                if (null === $mediafolder = $em->find('BackBee\NestedNode\MediaFolder', md5('media'))) {
                    $mediafolder = new \BackBee\NestedNode\MediaFolder(md5('media'));
                    $mediafolder->setTitle('Mediacenter')->setUrl('/');
                    $em->persist($mediafolder);
                }
```

File: Console/Command/CreateWebsiteCommand.php [hunk 22]
```diff
@@ -0,0 +1,227 @@
...
+        return $this;
```
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: I think that `$this->entyMgr->flush()` is missing before to conclude.

Also, this is a huge method, could try to write some mote tiny specific method ?

File: Console/Command/CreateWebsiteCommand.php (hunk 21 above)
Comment (crouillon) [inline Console/Command/CreateWebsiteCommand.php]: And Keyword root on an empty db:
            ```
if (null === $em->find('BackBee\NestedNode\KeyWord', md5('root'))) {
                $keyword = new \BackBee\NestedNode\KeyWord(md5('root'));
                $keyword->setRoot($keyword);
                $keyword->setKeyWord('root');
                $em->persist($keyword);
            }
```

File: Console/Command/UpdateUsersRightsCommand.php [hunk 23]
```diff
@@ -0,0 +1,830 @@
...
+
+        $this->output = $output;
+        if ($input->getOption('verbose')) {
+            $this->output->setVerbosity(OutputInterface::VERBOSITY_VERBOSE);
+        } else {
+            $this->output->setVerbosity(OutputInterface::VERBOSITY_NORMAL);
+        }
+
+        $this->em->getConnection()->getConfiguration()->setSQLLogger(null);
+        $this->bbapp = $this->getContainer()->get('bbapp');
```
Comment (crouillon) [inline Console/Command/UpdateUsersRightsCommand.php]: same as $this->app already declared

File: Console/Command/ApplicationUpdateCommand.php (hunk 4 above)
Comment (oanesbogdan) [inline Console/Command/ApplicationUpdateCommand.php]: The command supports now these options: --host -port --user --password
The file doctrine.yml shouldn't be updated with the new values?

Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: No, the main aim for these options is to be able to use other db conf from the one stored in doctrine.yml

File: Console/Command/ApplicationUpdateCommand.php [hunk 24]
```diff
@@ -70,57 +84,136 @@ protected function configure()
...
-            $this->em->getConnection()->executeUpdate('SET FOREIGN_KEY_CHECKS=1');
+            if ($drop) {
+                $sqls = array_merge($schema->getDropDatabaseSQL(), $sqls);
+            }
+
+            if ($force) {
+                $output->writeln('<info>Running drop/update</info>');
+
+                $this->em->getConnection()->executeUpdate('SET FOREIGN_KEY_CHECKS=0');
+                $schema->dropDatabase();
```
Comment (crouillon) [inline Console/Command/ApplicationUpdateCommand.php]: Only if $drop is true

**Comment by crouillon** [comment]
:+1: Thanks @oanesbogdan

**Pull request merged**